allennlp/nn/util.py

```diff
@@ -14,6 +14,24 @@
 
 logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
+def move_to_device(obj, cuda_device: int):
+    """
+    Given a structure (possibly) containing Tensors on the CPU,
+    move all the Tensors to the specified GPU (or do nothing, if they should be on the CPU).
+    """
+    if cuda_device < 0:
+        return obj
+
+    if isinstance(obj, torch.Tensor):
+        return obj.cuda(cuda_device)
+    elif isinstance(obj, dict):
+        return {key: move_to_device(value, cuda_device) for key, value in obj.items()}
+    elif isinstance(obj, list):
+        return [move_to_device(item, cuda_device) for item in obj]
```

Comment:
I don't know if fixing this is worth it or not, but these three lower blocks all create new objects.  If we're trying to minimize the amount of computation done on the CPU inside the training loop, we probably want to avoid creating these objects (though, I'm making this point without any profiling, so I could just be wrong about whether this is actually an issue).

I mean, imagine what happens for the input to the dependency parser.  There are some tensors, and they get moved immediately and quickly.  There are also some metadata fields that are just intended to be passed through to the demo.  All of those essentially get a _deepcopy_ here, just for fun.  And this is pretty typical for our models, because the way that we've written the demo code means that you need to pass through these metadata fields.

So, here's one more proposed test for this before merging: actually run training on the GPU for the dependency parser, or BiDAF, or something, before and after this PR, and see if running time is affected.  It'd also be nice to try to a third variant, where we do something like `if not has_tensor(obj): return obj`, just returning the object directly if there's no tensor.  You could go crazy with a fourth variant and try to just _modify_ the object with CUDA tensors instead of creating a new one, but I'd guess that case is not really worth the effort, as objects where that happens tend to be small (the biggest one I can think of is the semantic parsing code, where we have large lists of `ProductionRuleArrays`, which are complex structures that have tensors inside).

What do you think?  Am I just being crazy here?

Reply:
my intuition is that `deepcopy` is a very fast operation (unless at some point you encounter an object that's intrinsically expensive to copy, which I don't think is the case here), but I can test this

Reply:
this is on bidaf

create tensors on GPU:

```
start_acc: 0.3215, end_acc: 0.3514, span_acc: 0.2418, em: 0.2702, f1: 0.3893, loss: 5.2269 ||: 100%|##########| 2190/2190 [23:22<00:00,  1.56it/s]
start_acc: 0.5415, end_acc: 0.5814, span_acc: 0.4426, em: 0.5521, f1: 0.6738, loss: 3.1927 ||: 100%|##########| 265/265 [01:12<00:00,  3.66it/s]
start_acc: 0.5324, end_acc: 0.5789, span_acc: 0.4346, em: 0.4852, f1: 0.6342, loss: 3.2254 ||: 100%|##########| 2190/2190 [19:34<00:00,  1.86it/s]
start_acc: 0.5861, end_acc: 0.6245, span_acc: 0.4947, em: 0.6122, f1: 0.7212, loss: 2.7903 ||: 100%|##########| 265/265 [00:45<00:00,  5.80it/s]
start_acc: 0.5793, end_acc: 0.6252, span_acc: 0.4823, em: 0.5370, f1: 0.6832, loss: 2.8311 ||: 100%|##########| 2190/2190 [19:36<00:00,  1.86it/s]
start_acc: 0.6026, end_acc: 0.6440, span_acc: 0.5114, em: 0.6305, f1: 0.7376, loss: 2.6503 ||: 100%|##########| 265/265 [00:45<00:00,  5.80it/s]
```

create tensors on CPU, move to GPU

```
start_acc: 0.3212, end_acc: 0.3506, span_acc: 0.2404, em: 0.2689, f1: 0.3895, loss: 5.2259 ||: 100%|##########| 2190/2190 [24:08<00:00,  1.51it/s]
start_acc: 0.5368, end_acc: 0.5741, span_acc: 0.4410, em: 0.5530, f1: 0.6716, loss: 3.2478 ||: 100%|##########| 265/265 [01:19<00:00,  3.33it/s]
start_acc: 0.5336, end_acc: 0.5792, span_acc: 0.4356, em: 0.4866, f1: 0.6344, loss: 3.2282 ||: 100%|##########| 2190/2190 [20:19<00:00,  1.80it/s]
start_acc: 0.5898, end_acc: 0.6286, span_acc: 0.5013, em: 0.6130, f1: 0.7215, loss: 2.7526 ||: 100%|##########| 265/265 [00:51<00:00,  5.13it/s]
start_acc: 0.5808, end_acc: 0.6252, span_acc: 0.4822, em: 0.5375, f1: 0.6836, loss: 2.8326 ||: 100%|##########| 2190/2190 [20:20<00:00,  1.79it/s]
start_acc: 0.6047, end_acc: 0.6440, span_acc: 0.5150, em: 0.6348, f1: 0.7438, loss: 2.6429 ||: 100%|##########| 265/265 [00:53<00:00,  4.96it/s]
```

the time per training epoch increases from 19:30 to 20:20, but you'd expect it to increase a little bit. I haven't yet tried your third variant, my guess is that the extra `isinstance` checks make it more expensive than the deepcopy way, but we'll see.